### PR TITLE
ARROW-5082: [Python] Stop exporting copies of shared libraries in wheel

### PR DESCRIPTION
**Pull request by fsaintjacques**
This reduces the size from 50mb to 28mb. I haven't tested if this breaks the OSX wheel. Note that the fix is brittle since currently (on Linux) it links with the full-versioned shared library binary. This works out due to the `-len(x)` sorting applied. A proper fix would be to keep the one linked (found via `ldd` or some other methods). I suspect that auditwheel and subsequent install & test will catch this if the contract were to change.

### CODE REVIEW

File: python/manylinux1/build_arrow.sh
Comment (wesm) [inline python/manylinux1/build_arrow.sh]: Is there any documentation about this?

File: python/setup.py
Comment (wesm) [inline python/setup.py]: If the user wants to link a C++ library to the libarrow that is bundled with wheels, will it still work if there is only `libarrow.so.14` there?

**Comment by wesm** [comment]
@kszucs can we run the Crossbow wheel builds before merging this? Also want to wait for @xhochy to review

**Comment by kszucs** [comment]
@ursabot crossbow --help

**Comment by kszucs** [comment]
@ursabot crossbow package wheel

**Comment by kszucs** [comment]
Most of the wheel builds will fail mostly because of the missing OpenSSL dependency - so OpenSSL should be turned off.

**Comment by nealrichardson** [comment]
https://github.com/apache/arrow/pull/4494 may also fix the OpenSSL issue, if we can get that in.

**Comment by kszucs** [comment]
@nealrichardson You may try to execute the same ursabot command, see whether the PR fixes the wheel builds (execute `@ursabot crossbow package wheel conda linux` to run everything).

File: python/setup.py
Comment (xhochy) [reply python/setup.py]: Tensorflow is doing the same. Downstream users might need to add something to their linker commands though.

File: python/manylinux1/build_arrow.sh
Comment (kou) [reply python/manylinux1/build_arrow.sh]: pyarrow Debian package doesn't exist.
Source distribution is needed to use pyarrow with libarrow Debian package.

File: python/setup.py
Comment (fsaintjacques) [reply python/setup.py]: Side note, when I extracted the tensorflow packages, I noted they used the versionless path.

File: python/manylinux1/build_arrow.sh
Comment (fsaintjacques) [reply python/manylinux1/build_arrow.sh]: Could you care to explain, I'm missing some pieces. My comment was regarding uwe comment on zulip.

**Comment by wesm** [comment]
How important is it to have the shared library with the ABI version tag in these wheels? I see two options:

* Ship `libarrow.so.$ABI_VERSION`
* Ship unversioned `libarrow.so`

Either of these is OK with me. @xhochy do you have an opinion so we can get this closed out?

**Comment by xhochy** [comment]
I would prefer the ABI version in the name as this is the more graceful error for the end user than getting a segmentation fault. We should test first if users of the wheel (read turbodbc) need to adjust their build system or whether the ABI named version is still picked up

**Comment by wesm** [comment]
OK, who wants to do that? We're reaching the critical horizon for 0.14 so need to get this issue closed out and move on to the other backlog items

**Comment by fsaintjacques** [comment]
I think I've reached my limit of wheels debugging. @xhochy can you pickup the validation of turbobc?

File: python/manylinux1/build_arrow.sh
Comment (pitrou) [reply python/manylinux1/build_arrow.sh]: In any case, I'm not sure why this needs to be there. Anyone can build a source distribution from scratch (as the name suggests, it just packages source code together).

Comment (xhochy) [reply python/manylinux1/build_arrow.sh]: This gives us a reproducible and controlled environment to build a source tarball. You still need the right versions of setuptools amd friends to actually support Markdown in the description field of the package information.

**Comment by wesm** [comment]
I can take this for a spin tomorrow (Tuesday) if no one else volunteers since I have built turbodbc before

**Comment by wesm** [comment]
I've tested the wheel with turbodbc locally and turbodbc seems to be fine with linking to the versioned .so file

(note that it's saying "not found" because the pyarrow wheel directory is not in my LD_LIBRARY_PATH)

```
$ ldd turbodbc_arrow_support.cpython-37m-x86_64-linux-gnu.so 
	linux-vdso.so.1 (0x00007ffe59771000)
	libturbodbc.so => /home/wesm/code/turbodbc/build/libturbodbc.so (0x00007f72a07aa000)
	libarrow.so.14 => not found
	libarrow_python.so.14 => not found
	libcpp_odbc.so => /home/wesm/code/turbodbc/build/cpp/cpp_odbc/Library/libcpp_odbc.so (0x00007f72a0724000)
	libstdc++.so.6 => /home/wesm/cpp-runtime-toolchain/lib/libstdc++.so.6 (0x00007f72a05e2000)
	libgcc_s.so.1 => /home/wesm/cpp-runtime-toolchain/lib/libgcc_s.so.1 (0x00007f72a05cc000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f72a03e1000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f72a0293000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f72a0949000)
	libodbc.so.2 => /home/wesm/miniconda/envs/turbodbc-dev/lib/libodbc.so.2 (0x00007f72a0212000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f72a020c000)
	libiconv.so.2 => /home/wesm/miniconda/envs/turbodbc-dev/lib/./libiconv.so.2 (0x00007f72a0122000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f72a0101000)
```

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**
